jujugui/static/gui/src/app/app.js

```diff
@@ -660,6 +660,12 @@ YUI.add('juju-gui', function(Y) {
             this.controllerAPI.connect();
           }
         }
+        // If we have a jimmURL then we need to rely on the controllerAPI
```

Comment:
Ok so my impression is that the logic around the GUI connection is starting to be a bit complex.
My impression is that the GUI should not have a concept of jimm: a jimmURL config options smells.
The promise is that we can connect to a Juju controller or to a Jimm in a transparent way, without the client having to know something specific to the latter, i.e. without the client having any special logic related to jimm. The only exception is handling the redirections when logging in to models, but that can be handled generically, as a possibility in the API, without knowing there is a jimm.

I'd really like to see some simplifications and less special cases in the code: we already have 4 config values that we use to build the WebSocket URLs where to connect:
1) apiAddress from which we can parse the base host and port;
2) jujuEnvUUID that holds the UUID;
3) socketTemplate that we can use together with host, port and uuid from points 1 and 2 to build the URL for establishing the model connection;
4) controllerSocketTemplate that we can use together with host and port from point 1 to build the URL for establishing the controller connection.

So connecting to jimm from the GUI perspective is not a special case: it is just a controller connection, so point 4 should be used, no need for introducing a new config value.

Let's consider this code here:
```javascript
         if (isNotGISFSandbox || isNotConnected) {
           this.env.connect();
           if (this.controllerAPI) {
             // We won't have a controller API connection in Juju 1 or in
             // sandbox mode.
              this.controllerAPI.connect();
            }
          }
          if (this.get('jimmURL')) {
           this.controllerAPI.connect();
         }
```
In this snippet, `this.controllerAPI.connect();` could be called twice based on unrelated stuff. Moreover, I am not sure why `this.env.connect()` should not be called in sandbox mode, and it's surely wrong that in sandbox mode we don't have a controller connection. To be clear, IMHO we should remove all special cases for sandbox mode and for jimm.